packages/zoe/src/contractFacet/reallocate.js

```diff
@@ -0,0 +1,104 @@
+import { makeScalarMapStore } from '@agoric/vat-data';
+
+import { assertRightsConserved } from './rightsConservation.js';
+import { addToAllocation, subtractFromAllocation } from './allocationMath.js';
+
+const { Fail } = assert;
+
+/** @typedef {Array<AmountKeywordRecord>} TransactionList */
+
+/**
+ * Convert from a list of transfer descriptions ([fromSeat, toSeat, fromAmount,
+ * toAmount], with many parts optional) to a list of resulting allocations for
+ * each of the seats mentioned.
+ *
+ * @param {Array<TransferPart>} transfers
+ * @returns {[ZCFSeat,AmountKeywordRecord][]}
+ */
+export const makeAllocationMap = transfers => {
+  /** @type {MapStore<ZCFSeat, [TransactionList, TransactionList]>} */
+  const allocations = makeScalarMapStore();
+
+  const getAllocations = seat => {
+    if (allocations.has(seat)) {
+      return allocations.get(seat);
+    }
+
+    /** @type {[TransactionList, TransactionList]} */
+    const pair = [[], []];
+    allocations.init(seat, pair);
+    return pair;
+  };
+
+  const updateAllocations = (seat, newAllocation) => {
+    allocations.set(seat, newAllocation);
+  };
+
+  const decrementAllocation = (seat, decrement) => {
+    const [incr, decr] = getAllocations(seat);
+
+    const newDecr = [...decr, decrement];
+    updateAllocations(seat, [incr, newDecr]);
+  };
+
+  const incrementAllocation = (seat, increment) => {
+    const [incr, decr] = getAllocations(seat);
+
+    const newIncr = [...incr, increment];
+    updateAllocations(seat, [newIncr, decr]);
+  };
+
+  for (const [
+    fromSeat = undefined,
+    toSeat = undefined,
+    fromAmounts = undefined,
+    toAmounts = undefined,
+  ] of transfers) {
```

Comment:
non-blocking: the default values are redundant. they don't affect the runtime and as a matter of documentation it's already in the static type.
```suggestion
  for (const [fromSeat, toSeat, fromAmounts, toAmounts] of transfers) {
```

Reply:
done.